File: array_partition_lc561.py
"""
Author: Srinidhi
Time Complxity: O(N log N) - Sorting
Space Complexity: O(1) - Constant space for variables
Did it run on LC: yes

Logic: Thinking about this in greedy way, to get the max overall,
we need to sort it, so that when numbers are paired, the max and max-1
dont get comprimised
"""
class Solution:
    def arrayPairSum(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        
        if len(nums)==2:
            return min(nums)
        
        curr_sum = 0
        nums.sort()
        for i in range(0,len(nums)-1,2):
            curr_sum = curr_sum + min(nums[i],nums[i+1])
        return curr_sum

File: ArrayPartitionLC561.java
class Solution {

    //Time Complexity: O(n*log(n))
    //Space Complexity: O(1)

    public int arrayPairSum(int[] nums) {

        if(nums == null || nums.length == 0) return 0;                  //check for null case
        
        Arrays.sort(nums);                                  //sort the array
        
        int sum = 0;                                //store the sum

        for(int i=0; i<nums.length; i+=2){          //iterate through nums array, increament the i by 2
            sum += nums[i];                         //getting the eleement from num and add it into the sum
        }
        return sum;                                 //return sum
    }
}
